File: packages/nvidia-prefetch/src/nix-file.ts
import { $ } from 'bun';
import { logger, execSafe, escapeNixString, validateNixFile, findRepoRoot } from '@euvlok/shared';
import { join } from 'pathe';

async function find(): Promise<string | null> {
  const root = await findRepoRoot();
  if (!root) return null;

  const result = await execSafe([
    'find',
    root,
    '-name',
    'nvidia-driver.nix',
    '-type',
    'f',
    '-not',
    '-path',
    '*/.git/*',
    '-not',
    '-path',
    '*/.jj/*',
    '-not',
    '-path',
    '*/node_modules/*',
  ]);

  if (result.exitCode !== 0 || !result.stdout) return null;
  const path = result.stdout.split('\n')[0];
  return path && (await Bun.file(path).exists()) ? path : null;
}

export async function getCurrentVersion(): Promise<string | null> {
  const path = await find();
  if (!path) return null;

  const content = await Bun.file(path).text();
  const match = content.match(/^\s*version\s*=\s*"([^"]+)"/m);
  return match ? match[1] : null;
}

export async function updateNvidiaDriverNix(
  version: string,
  sha256_64bit: string,
  sha256_aarch64: string,
  openSha256: string,
  settingsSha256: string,
  persistencedSha256: string,
): Promise<void> {
  const path = await find();
  if (!path) {
    throw new Error(
      "Could not find modules/nixos/nvidia-driver.nix file. Make sure you're running this script from within the repository",
    );
  }

  logger.info(`Updating ${path}...`);

  const backup = `${path}.backup`;
  await Bun.write(backup, Bun.file(path));
  logger.info(`Created backup: ${backup}`);

  const content = `{
  version = "${escapeNixString(version)}";
  sha256_64bit = "${escapeNixString(sha256_64bit)}";
  sha256_aarch64 = "${escapeNixString(sha256_aarch64)}";
  openSha256 = "${escapeNixString(openSha256)}";
  settingsSha256 = "${escapeNixString(settingsSha256)}";
  persistencedSha256 = "${escapeNixString(persistencedSha256)}";
}
`;

  const tmp = join(Bun.env.TMPDIR || '/tmp', `nvidia-driver-${Date.now()}.nix`);
  await Bun.write(tmp, content);

  try {
    await validateNixFile(tmp);
  } catch {
    await Bun.write(path, Bun.file(backup));
    await $`rm -f ${tmp}`.quiet();
    await $`rm -f ${backup}`.quiet();
    throw new Error('Generated nix file is invalid, restored backup');
  }

  await Bun.write(path, Bun.file(tmp));
  await $`rm -f ${tmp}`.quiet();
  await $`rm -f ${backup}`.quiet();

  logger.success(`Successfully updated ${path}`);
}

File: packages/nvidia-prefetch/src/nix_file.zig
const std = @import("std");
const DriverHashes = @import("hash.zig").DriverHashes;

pub fn currentVersion(allocator: std.mem.Allocator, io: std.Io) !?[]const u8 {
    const path = try find(allocator, io) orelse return null;
    defer allocator.free(path);

    const content = try std.Io.Dir.cwd().readFileAlloc(io, path, allocator, .limited(1024 * 1024));
    defer allocator.free(content);

    return extractStringValue(allocator, content, "version");
}

pub fn update(allocator: std.mem.Allocator, io: std.Io, driver_version: []const u8, hashes: DriverHashes) !void {
    const path = try find(allocator, io) orelse return error.NvidiaDriverNixNotFound;
    defer allocator.free(path);

    std.debug.print("info: Updating {s}...\n", .{path});

    const content = try format(allocator, driver_version, hashes);
    defer allocator.free(content);

    try std.Io.Dir.cwd().writeFile(io, .{ .sub_path = path, .data = content });
    std.debug.print("success: Successfully updated {s}\n", .{path});
}

fn find(allocator: std.mem.Allocator, io: std.Io) !?[]const u8 {
    const cwd = try std.process.currentPathAlloc(io, allocator);
    defer allocator.free(cwd);

    var cursor: []const u8 = cwd;
    while (true) {
        const candidate = try std.fs.path.join(allocator, &.{ cursor, "modules/nixos/nvidia-driver.nix" });
        std.Io.Dir.cwd().access(io, candidate, .{}) catch |err| {
            allocator.free(candidate);
            switch (err) {
                error.FileNotFound => {},
                else => |e| return e,
            }
            const parent = std.fs.path.dirname(cursor) orelse return null;
            if (std.mem.eql(u8, parent, cursor)) return null;
            cursor = parent;
            continue;
        };
        return candidate;
    }
}

fn extractStringValue(allocator: std.mem.Allocator, content: []const u8, name: []const u8) !?[]const u8 {
    var lines = std.mem.splitScalar(u8, content, '\n');
    while (lines.next()) |line| {
        const trimmed = std.mem.trim(u8, line, " \t");
        if (!std.mem.startsWith(u8, trimmed, name)) continue;

        var rest = std.mem.trim(u8, trimmed[name.len..], " \t");
        if (rest.len == 0 or rest[0] != '=') continue;

        rest = std.mem.trim(u8, rest[1..], " \t");
        if (rest.len == 0 or rest[0] != '"') continue;

        const end = std.mem.indexOfScalarPos(u8, rest, 1, '"') orelse continue;
        return try allocator.dupe(u8, rest[1..end]);
    }

    return null;
}

fn format(allocator: std.mem.Allocator, driver_version: []const u8, hashes: DriverHashes) ![]const u8 {
    var out: std.Io.Writer.Allocating = .init(allocator);
    defer out.deinit();

    try out.writer.writeAll("{\n");
    try writeString(&out.writer, "version", driver_version);
    try writeString(&out.writer, "sha256_64bit", hashes.sha256);
    try writeString(&out.writer, "sha256_aarch64", hashes.sha256_aarch64);
    try writeString(&out.writer, "openSha256", hashes.openSha256);
    try writeString(&out.writer, "settingsSha256", hashes.settingsSha256);
    try writeString(&out.writer, "persistencedSha256", hashes.persistencedSha256);
    try out.writer.writeAll("}\n");
    return out.toOwnedSlice();
}

fn writeString(writer: *std.Io.Writer, name: []const u8, value: []const u8) !void {
    try writer.print("  {s} = \"", .{name});
    for (value) |byte| switch (byte) {
        '\\', '"' => try writer.print("\\{c}", .{byte}),
        '\n' => try writer.writeAll("\\n"),
        '\r' => try writer.writeAll("\\r"),
        '\t' => try writer.writeAll("\\t"),
        else => try writer.writeByte(byte),
    };
    try writer.writeAll("\";\n");
}
